### PR TITLE
App part 1 

### LINKED ISSUE #89: Form part 1 (basic hacker info) with no state and validation
Basic info for hackers
- [x] add form input components with no state management
- [x] no need to include header, progress bar, next/back buttons at the bottom
- [x] add form state

Screen Shot 2020-11-13 at 5 07 39 PM

### PR DESCRIPTION
**Pull request by shuting-hu**
Closes #89 
- form skeleton with inputs and local states
- new components for centering H1 and form spacing between questions
- refactoring, new app form folder in components to separate presentation from logic in containers
- textinput/select/dropdown have inconsistent margins so alignment is off (being fixed in separate PR)
- autosave to db, error validation, navigation buttons left for #112 

https://drive.google.com/file/d/1xLGEnYYdPoG5qQZ2qDYStlu3QHctZec_/view?usp=sharing
gif was too big boooo

### CODE REVIEW

**Comment by ianmah** [comment]
Containers should handle as much logic as possible, and components should handle as much presentation as much as possible. The container you've made for Part1 handles just a little bit too much of the presentation. Take a look at this component for an idea on how to separate the presentation from the logic. https://github.com/nwplus/livesite/blob/master/src/components/ProjectForm.js

File: src/pages/Charcuterie.js [hunk 1]
```diff
@@ -77,7 +77,7 @@ export default () => {
         <br></br>
         <QuestionHeading>this is a question heading</QuestionHeading>
         <QuestionHeading>question 14</QuestionHeading>
-        <H1>How did you hear about nwHacks?</H1>
+        <H1 size="1.5em">How did you hear about nwHacks?</H1>
```
Comment (jackyzha0) [inline src/pages/Charcuterie.js:80]: if we need a larger H1 for application questions, we can just create a styled H1 like so

```jsx
const QuestionSubheading = styled(H1)`
	size: 1.5em;
`
```

File: src/containers/Application/Part1.js [hunk 2]
```diff
@@ -1,6 +1,424 @@
...
+  { value: 2019, label: '2019' },
+  { value: 2020, label: '2020' },
+  { value: 2021, label: '2021' },
+  { value: 2022, label: '2022' },
+  { value: 2023, label: '2023' },
+  { value: 2024, label: '2024' },
+  { value: 2025, label: '2025+' },
+]
+
+// check w logs
```
Comment (jackyzha0) [inline src/containers/Application/Part1.js:34]: might be good to just do 0, 1, 2, 3, 4+ cause anything past that is pretty much the same :angrycry:

File: src/pages/Charcuterie.js (hunk 1 above)
Comment (shuting-hu) [reply src/pages/Charcuterie.js:80]: I initially made this exact component but then noticed H1 had a size conditional so I used that instead, can bring the styled H1 component back though :))

File: src/containers/Application/Part1.js (hunk 2 above)
Comment (shuting-hu) [reply src/containers/Application/Part1.js:34]: roger that!

File: src/containers/Application/Part1.js [hunk 3]
```diff
@@ -1,6 +1,424 @@
...
+  { value: 5, label: '5' },
+  { value: 6, label: '6' },
+  { value: 7, label: '7' },
+  { value: 8, label: '8' },
+  { value: 9, label: '9' },
+  { value: 10, label: '10+' },
+]
 
 // form part 1
 export default ({ children }) => {
```
Comment (panjenny0) [inline src/containers/Application/Part1.js:50]:
```suggestion
export default () => {
```

**Review by panjenny0 — CHANGES_REQUESTED** [review]
Great job on the form! To separate _what_ should be rendered from _how_ it should be rendered, we should to put the form components in a new file in `components` 
```
+-- ApplicationForm     <-- Create a new folder
|   +-- index.js        <-- what used to be ApplicationForm.js
|   +-- BasicInfo.js    <-- Form part 1
|   +-- Skills.js        <-- Form part 2
|   +-- Questionnaire.js    <-- Form part 3
```
Eventually, we're gonna need to fetch the user's existing inputs in `container/Application/Part1.js` so we need to be able to pass in form values into the form components.
### I would do this:
- everything except for your states (i.e. all the options, the imported JSONs, just not the useState) can go inside `BasicInfo.js`
```
<BasicInfo formInputs={states} onChange={setStates} />
```
- `BasicInfo.js` would take in your input states and onChange
```
export default ({ formInputs, onChange }) => {
```
- modify how your components are rendered in `BasicInfo.js` so that they can take in form values, for example:
```
<TextInput
          placeholder="First Name"
          value={formInputs.firstName}
          onChange={e =>
            onChange({
              ...states,
              firstName: e.target.value,
            })
          }
        ></TextInput>
```

For comparison, take a look at `ProjectForm.js`
```
<TextInput
            value={project.devpostUrl}
            placeholder="Required"
            onChange={e => onChange({ ...project, devpostUrl: e.target.value })}
          />
```
FYI @kevin-zou @anneguo3 let's follow the same convention for parts 2 and 3 of the form

File: src/components/ApplicationForm/BasicInfo.js [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+        type="checkbox"
+        label="Asian"
+        checked={formInputs.ethnicity.asian}
+        onChange={() =>
+          onChange({
+            ...formInputs,
+            ethnicity: { ...formInputs.ethnicity, asian: !formInputs.ethnicity.asian },
+          })
+        }
+      ></Select>
```
Comment (ianmah) [inline src/components/ApplicationForm/BasicInfo.js:113]: Can we use an map of options like you did for graduation? Also for the options for the "contribution" question (q11)

File: src/containers/Application/Part1.js [hunk 5]
```diff
@@ -1,6 +1,37 @@
...
+    },
+    isOfLegalAge: null,
+    phoneNumber: '',
+    school: '',
+    major: '',
+    educationLevel: '',
+    graduation: 0,
+    hackathonsAttended: 0,
+    contributionRole: '',
+    location: '',
```
Comment (ianmah) [inline src/containers/Application/Part1.js:29]: Do we need to declare all these as empty? Can we init it as an empty object

Comment (Ciuffi) [reply src/containers/Application/Part1.js:29]: Might be better to replace this with `updateHackerApplication` once its pushed to master

**Review by panjenny0 — CHANGES_REQUESTED** [review]
Some final changes!

Also when updating your branch with the latest version of master, theres some conflicts you might have to fix in `Form.js` - all you need to do is remove the `FormContainer` inport and remove the `FormContainer` wrappers
After updating your branch, you should be able to go to `/application/part-1` to see your form :)

File: src/components/ApplicationForm/BasicInfo.js [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        checked={formInputs.contributionRole === 'Business Strategist'}
+        onChange={() => onChange({ ...formInputs, contributionRole: 'Business Strategist' })}
+      ></Select>
+      <Select
+        type="radio"
+        label="Other"
+        checked={formInputs.contributionRole === 'Other'}
+        onChange={() => onChange({ ...formInputs, contributionRole: 'Other' })}
+      ></Select>
+      {/* insert new input component for 'Please Specify' */}
```
[the remark this answers is not in the document]
Comment (shuting-hu) [reply src/components/ApplicationForm/BasicInfo.js:390]: logs no longer wants other options but @kevin-zou can use the 'please specify' component for your dropdown other in part 3!

File: src/components/ApplicationForm/BasicInfo.js [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+]
+
+// check w logs
+const hackathonOptions = [
+  { value: 0, label: '0' },
+  { value: 1, label: '1' },
+  { value: 2, label: '2' },
+  { value: 3, label: '3' },
+  { value: 4, label: '4+' },
+]
```
[the remark this answers is not in the document]
Comment (shuting-hu) [reply src/components/ApplicationForm/BasicInfo.js:43]: just slacked JP, nwHacks logs approves of these options! 🎉

File: src/components/ApplicationForm/BasicInfo.js [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+      ></Select>
+    </FormSpacing>
+
+    <FormSpacing>
+      <QuestionHeading>question 04</QuestionHeading>
+      <H1 size="1.5em">Will you be 19 years or older by January 9th, 2021?</H1>
+      <Select
+        type="radio"
+        label="Yes"
+        checked={formInputs.isOfLegalAge === true}
```
Comment (panjenny0) [inline src/components/ApplicationForm/BasicInfo.js:217]:
```suggestion
        checked={formInputs.isOfLegalAge}
```

File: src/components/ApplicationForm/BasicInfo.js [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+        onChange={() => onChange({ ...formInputs, isOfLegalAge: true })}
+      ></Select>
+      <Select
+        type="radio"
+        label="No"
+        checked={formInputs.isOfLegalAge === false}
```
Comment (panjenny0) [inline src/components/ApplicationForm/BasicInfo.js:223]:
```suggestion
        checked={!formInputs.isOfLegalAge}
```

File: src/components/ApplicationForm/BasicInfo.js [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+        type="radio"
+        label="Developer"
+        checked={formInputs.contributionRole === 'Developer'}
+        onChange={() => onChange({ ...formInputs, contributionRole: 'Developer' })}
+      ></Select>
+      <Select
+        type="radio"
+        label="Designer"
+        checked={formInputs.contributionRole === 'Designer'}
+        onChange={() => onChange({ ...formInputs, contributionRole: 'Designer' })}
```
Comment (panjenny0) [inline src/components/ApplicationForm/BasicInfo.js:356]: nit: can these be lowercase?
```suggestion
        checked={formInputs.contributionRole === 'developer'}
        onChange={() => onChange({ ...formInputs, contributionRole: 'developer' })}
      ></Select>
      <Select
        type="radio"
        label="Designer"
        checked={formInputs.contributionRole === 'designer'}
        onChange={() => onChange({ ...formInputs, contributionRole: 'designer' })}
```

**Review by panjenny0 — COMMENTED** [review]
some nit things then should be good to merge!!

**Review by panjenny0 — DISMISSED** [review]

File: src/containers/Application/Part1.js (hunk 5 above)
Comment (panjenny0) [reply src/containers/Application/Part1.js:29]: I think initializing this is fine - we need a way to set the initial values if it's a new application

File: src/components/ApplicationForm/BasicInfo.js (hunk 4 above)
Comment (shuting-hu) [reply src/components/ApplicationForm/BasicInfo.js:113]: these are multiselects, how should I map it?

**Review by panjenny0 — DISMISSED** [review]

File: src/components/ApplicationForm/BasicInfo.js (hunk 4 above)
Comment (ianmah) [reply src/components/ApplicationForm/BasicInfo.js:113]: wrote this on github idk if the syntax is correct but something like this
```
const ethnicityOptions = {
    asian: "Asian",
    black: "Black or African American"
}
{
Object.entries(ethnicityOptions).map(([ethnicity, label]) => {
    return 
    <Select
        type="checkbox"
        label={label}
        checked={formInputs.ethnicity[ethnicity]}
        onChange={() =>
          onChange({
            ...formInputs,
            ethnicity: { ...formInputs.ethnicity, [ethnicity]: !formInputs.ethnicity[ethnicity] },
          })
        }
      />
}
}
```

Comment (shuting-hu) [reply src/components/ApplicationForm/BasicInfo.js:113]: wouldn't this only make it one select button?

**Review by ianmah — APPROVED** [review]
✨

**Comment by acchiang** [comment]
@shuting-hu quicc! merge before the conflicts start appearing

**Review by panjenny0 — APPROVED** [review]
güd work 🚀 !!!

**Pull request merged**